### PR TITLE
improvement: Add step-wide argument transforms.

### PR DESCRIPTION
**Pull request by jimsynz**

### CODE REVIEW

File: lib/reactor/builder.ex [hunk 1]
```diff
@@ -21,15 +21,26 @@ defmodule Reactor.Builder do
...
+  @type step_options :: [async? | max_retries() | arguments_transform | context]
 
   @typedoc "Should the step be run asynchronously?"
   @type async? :: {:async?, boolean}
 
   @typedoc "How many times is the step allowed to retry?"
   @type max_retries :: {:max_retries, :infinity | non_neg_integer()}
 
+  @typedoc "Optionally transform all the arguments into new arguments"
+  @type arguments_transform :: {:transform, nil | (any -> map) | {module | keyword} | mfa}
```
Comment (zachdaniel) [inline lib/reactor/builder.ex:39]: Should this be `(map -> map)` ?

File: lib/reactor/builder.ex [hunk 2]
```diff
@@ -256,16 +289,33 @@ defmodule Reactor.Builder do
...
+  defp maybe_build_step_transform_step(arguments, name, transform)
+       when is_function(transform, 1),
+       do: maybe_build_step_transform_step(arguments, name, {Step.TransformAll, fun: transform})
+
+  defp maybe_build_step_transform_step(arguments, name, transform) do
+    step = %Step{
+      arguments: arguments,
+      async?: true,
+      impl: transform,
+      name: {:transform, :for, name},
```
Comment (zachdaniel) [inline lib/reactor/builder.ex:303]: I think these builtin step names might potentially need some "identifier that people would never use" that we can document should not be used. i.e `{:__reactor__, ...}`

**Review by zachdaniel — COMMENTED** [review]
had some comments, but LGTM

File: lib/reactor/builder.ex (hunk 1 above)
Comment (jimsynz) [reply lib/reactor/builder.ex:39]: Yes.

**Pull request merged**